packages/common-substrate/src/project/versioned/ProjectManifestVersioned.ts

```diff
@@ -1,32 +1,44 @@
-// Copyright 2020-2022 OnFinality Limited authors & contributors
+// Copyright 2020-2021 OnFinality Limited authors & contributors
 // SPDX-License-Identifier: Apache-2.0
 
 import {SubqlDatasource} from '@subql/types';
 import {plainToClass} from 'class-transformer';
-import {IProjectManifest} from '../types';
+import {ISubstrateProjectManifest} from '../types';
 import {ProjectManifestV0_0_1Impl} from './v0_0_1';
 import {ProjectManifestV0_2_0Impl} from './v0_2_0';
+import {ProjectManifestV0_2_1Impl} from './v0_2_1';
+import {ProjectManifestV0_3_0Impl} from './v0_3_0';
 
 export type VersionedProjectManifest = {specVersion: string};
 
 const SUPPORTED_VERSIONS = {
   '0.0.1': ProjectManifestV0_0_1Impl,
   '0.2.0': ProjectManifestV0_2_0Impl,
+  '0.2.1': ProjectManifestV0_2_1Impl,
+  '0.3.0': ProjectManifestV0_3_0Impl,
 };
 
 type Versions = keyof typeof SUPPORTED_VERSIONS;
 
 type ProjectManifestImpls = InstanceType<typeof SUPPORTED_VERSIONS[Versions]>;
 
-export function manifestIsV0_0_1(manifest: IProjectManifest): manifest is ProjectManifestV0_0_1Impl {
+export function manifestIsV0_0_1(manifest: ISubstrateProjectManifest): manifest is ProjectManifestV0_0_1Impl {
   return manifest.specVersion === '0.0.1';
 }
 
-export function manifestIsV0_2_0(manifest: IProjectManifest): manifest is ProjectManifestV0_2_0Impl {
+export function manifestIsV0_2_0(manifest: ISubstrateProjectManifest): manifest is ProjectManifestV0_2_0Impl {
   return manifest.specVersion === '0.2.0';
 }
 
-export class ProjectManifestVersioned implements IProjectManifest {
+export function manifestIsV0_2_1(manifest: ISubstrateProjectManifest): manifest is ProjectManifestV0_2_1Impl {
+  return manifest.specVersion === '0.2.1';
+}
+
+export function manifestIsV0_3_0(manifest: ISubstrateProjectManifest): manifest is ProjectManifestV0_3_0Impl {
+  return manifest.specVersion === '0.3.0';
+}
+
+export class ProjectManifestVersioned implements ISubstrateProjectManifest {
```

Comment:
I think we can have a base class for this, the terra implementation is almost identical

Reply:
most of the identical functions uses `_impl` which is of type `ProjectManifestImpls` . How shall I create this type in base class?
https://github.com/DeveloperInProgress/subql/blob/472189516e3b8ac301f36f8e4cc10e7471677ef8/packages/common-substrate/src/project/versioned/ProjectManifestVersioned.ts#L23

Reply:
I don't think so, you can use generics and either have it extend the base class or the interface